
#Note: This code file assumes that you have already read in the functions in the "Example_MetaAnalysis_GemmaOutput_StreamlinedReadingInAStudy.R"
#.... but I also discovered a bug in that file that was making one of the FilteringDEResults_GoodAnnotation function choke on datasets that didn't include pipes (|) in their gene symbol annotation.
#.... so before running this code, I recommend going to that file and reading in the (now fixed) FilteringDEResults_GoodAnnotation function

#Megan Hagenauer
#2022-08-02

###################

#Reading in one of the datasets that is completely lacking gene symbol or gene ID annotation:
#This one is for Annaka's meta-analysis: GSE81024

#Before starting, I looked at the Gemma's page for GSE81024:
https://gemma.msl.ubc.ca/expressionExperiment/showExpressionExperiment.html?id=13018
#and found the ID for the platform - in this case the platform was an Agilent microarray chip, and the ID for that platform is GPL7202

#If I type "GPL7202" into Google, the first thing that pops up is GEO's Accession viewer:
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL7202
#This page contains the annotation for the probes provided by the Agilent company
#If I scroll down to the bottom of this page there is a data table with the gene annotation for each probe.
#Under that table is the option "Download full table" - I download it to the same folder where I am storing the differential expression results for GSE81024.

#Alright, the first part of the code is the same as what we did for the other datasets:

#I set the working directory to where I have the files for GSE81024:
setwd("~/Documents/SideProjects/BrainGMT/Gemma/Hippocampus/Download_20220728/13018_GSE81024_diffExpAnalysis_99567")

#I determine which files are there:
list.files()
# [1] "analysis.results.txt"        "GPL7202-9760.txt"            "resultset_ID480667.data.txt"
# [4] "resultset_ID480669.data.txt"

#And then I plug the file names for the resultsets into the ReadingInGemmaDE function:
ReadingInGemmaDE(ResultSetFileNames=c("resultset_ID480669.data.txt", "resultset_ID480667.data.txt"))

str(TempResultsJoined)
# 'data.frame':	41253 obs. of  13 variables:
# $ Element_Name                 : chr  "A_51_P470981" "A_51_P269703" "A_52_P529098" "A_52_P214058" ...
# $ Gene_Symbol                  : logi  NA NA NA NA NA NA ...
# $ Gene_Name                    : logi  NA NA NA NA NA NA ...
# $ NCBI_ID                      : logi  NA NA NA NA NA NA ...
# $ FoldChange_24.h              : num  -0.02049 0.0149 0.00917 -0.05741 -0.06237 ...
# $ Tstat_24.h                   : num  -2.115 1.33 0.761 -6.42 -4.13 ...
# $ PValue_24.h                  : num  0.06739 0.2203 0.4683 0.000205 0.003297 ...
# $ Gene_Symbol                  : logi  NA NA NA NA NA NA ...
# $ Gene_Name                    : logi  NA NA NA NA NA NA ...
# $ NCBI_ID                      : logi  NA NA NA NA NA NA ...
# $ FoldChange_lipopolysaccharide: num  -0.0441 0.0192 0.0194 -0.0578 0.1146 ...
# $ Tstat_lipopolysaccharide     : num  -4.55 1.72 1.61 -6.46 7.59 ...
# $ PValue_lipopolysaccharide    : num  1.87e-03 1.24e-01 1.46e-01 1.96e-04 6.37e-05 ...


#This is where things differ: 

#Now I read in the additional annotation for the Agilent microarray chip that we got off the GEO website:
MoreAnnotation<-read.delim("GPL7202-9760.txt", sep="\t", header=TRUE, stringsAsFactors = FALSE, comment.char = "#")

str(MoreAnnotation)
# 'data.frame':	41282 obs. of  17 variables:
# $ ID                  : chr  "A_51_P100021" "A_51_P100034" "A_51_P100052" "A_51_P100063" ...
# $ SPOT_ID             : chr  "A_51_P100021" "A_51_P100034" "A_51_P100052" "A_51_P100063" ...
# $ CONTROL_TYPE        : chr  "FALSE" "FALSE" "FALSE" "FALSE" ...
# $ REFSEQ              : chr  "" "NM_027162" "NM_198863" "NM_010727" ...
# $ GB_ACC              : chr  "AY454345" "NM_027162" "NM_198863" "NM_010727" ...
# $ GENE                : int  16656 69674 245450 16924 NA 69886 70285 17427 26896 171232 ...
# $ GENE_SYMBOL         : chr  "Hivep3" "Mif4gd" "Slitrk2" "Lnx1" ...
# $ GENE_NAME           : chr  "human immunodeficiency virus type I enhancer binding protein 3" "MIF4G domain containing" "SLIT and NTRK-like family, member 2" "ligand of numb-protein X 1" ...
# $ UNIGENE_ID          : chr  "Mm.302758" "Mm.390387" "Mm.336081" "Mm.440403" ...
# $ ENSEMBL_ID          : chr  "" "" "ENSMUST00000036043" "" ...
# $ TIGR_ID             : logi  NA NA NA NA NA NA ...
# $ ACCESSION_STRING    : chr  "gb|AY454345|tc|TC1584815|nap|NAP057482-1" "ref|NM_027162|ref|NM_001243586|ref|NM_001243587|ref|NM_001243584" "ref|NM_198863|ref|NM_001161431|ens|ENSMUST00000036043|ens|ENSMUST00000166241" "ref|NM_010727|ref|NM_001159577|ref|NM_001159579|ref|NM_001159580" ...
# $ CHROMOSOMAL_LOCATION: chr  "chr4:119807944-119808005" "chr11:115469328-115469269" "chrX:63908145-63908204" "chr5:74993810-74993386" ...
# $ CYTOBAND            : chr  "mm|4qD2.1" "mm|11qE2" "mm|XqA7.1" "mm|5qC3.3" ...
# $ DESCRIPTION         : chr  "Mus musculus clone pZAS3 328-2275 zinc finger protein ZAS3 (Krc) mRNA, partial cds; alternatively spliced. [AY454345]" "Mus musculus MIF4G domain containing (Mif4gd), transcript variant 1, mRNA [NM_027162]" "Mus musculus SLIT and NTRK-like family, member 2 (Slitrk2), transcript variant 2, mRNA [NM_198863]" "Mus musculus ligand of numb-protein X 1 (Lnx1), transcript variant 2, mRNA [NM_010727]" ...
# $ GO_ID               : chr  "GO:0003676(nucleic acid binding)|GO:0003677(DNA binding)|GO:0005622(intracellular)|GO:0005634(nucleus)|GO:00057"| __truncated__ "GO:0005488(binding)|GO:0005575(cellular_component)|GO:0005634(nucleus)|GO:0005737(cytoplasm)|GO:0006417(regulat"| __truncated__ "GO:0003674(molecular_function)|GO:0007409(axonogenesis)|GO:0016020(membrane)|GO:0016021(integral to membrane)" "GO:0004842(ubiquitin-protein ligase activity)|GO:0005515(protein binding)|GO:0005737(cytoplasm)|GO:0006511(ubiq"| __truncated__ ...
# $ SEQUENCE            : chr  "CATGGCTGGATTAACGTATGTGTGTGGTATATAGATACACAGAGAGAAACCAAAGTGGTG" "GAGACTTTTGTGGAGGAAGCCTGTTTCCTCCAGTCATGAGTGACTGCCTCACCAGGTTGG" "CTAAATGTGAATTGCCAAGAAAGGAAGTTCACTAACATCTCTGACCTACAGCCCAAACCT" "GAAGAATCAGATGTGGTGACATTCTTCTCGCTGTCAACGGTAGAAGTACATCGGGTATGA" ...

colnames(MoreAnnotation)
# [1] "ID"                   "SPOT_ID"              "CONTROL_TYPE"         "REFSEQ"              
# [5] "GB_ACC"               "GENE"                 "GENE_SYMBOL"          "GENE_NAME"           
# [9] "UNIGENE_ID"           "ENSEMBL_ID"           "TIGR_ID"              "ACCESSION_STRING"    
# [13] "CHROMOSOMAL_LOCATION" "CYTOBAND"             "DESCRIPTION"          "GO_ID"               
# [17] "SEQUENCE"  

#I pull out just the columns for the probe ID, gene symbol, and gene name (and Entrez or NCBI ID, if available) and stash them in a new, smaller data frame:
NeededAnnotation<-MoreAnnotation[,c(1,7:8)]


str(NeededAnnotation)
# 'data.frame':	41282 obs. of  3 variables:
#   $ ID         : chr  "A_51_P100021" "A_51_P100034" "A_51_P100052" "A_51_P100063" ...
# $ GENE_SYMBOL: chr  "Hivep3" "Mif4gd" "Slitrk2" "Lnx1" ...
# $ GENE_NAME  : chr  "human immunodeficiency virus type I enhancer binding protein 3" "MIF4G domain containing" "SLIT and NTRK-like family, member 2" "ligand of numb-protein X 1" ...

#In order to join our differential expression results with the additional annotation, I need to tell R that they each have a column that contains the probe IDs. To do this, that column needs to be named the same thing within both our differential expression results (TempResultsJoined) and in the additional annotation (NeededAnnotation)
#Within TempResultsJoined that column is called "Element_Name"
colnames(TempResultsJoined)
# [1] "Element_Name"                  "Gene_Symbol"                   "Gene_Name"                    
# [4] "NCBI_ID"                       "FoldChange_24.h"               "Tstat_24.h"                   
# [7] "PValue_24.h"                   "Gene_Symbol"                   "Gene_Name"                    
# [10] "NCBI_ID"                       "FoldChange_lipopolysaccharide" "Tstat_lipopolysaccharide"     
# [13] "PValue_lipopolysaccharide"   

#So I re-name the columns in NeededAnnotation so that probe ID is named "Element_Name". I also re-name the Gene Symbol and Gene Name columns because I want the new additional annotation to be obvious after we join the data-frames.
colnames(NeededAnnotation)<-c("Element_Name", "Gene_Symbol_New","Gene_Name_New")

#I load the library that contains the join function:
library(plyr)

#And join the two data frames so that the resulting data frame contains the same rows & order as our differential expression results (TempResultsJoined) using the "type=left" parameter:
TempResultsJoined_wAnnotation<-join(TempResultsJoined, NeededAnnotation, by="Element_Name", type="left")

str(TempResultsJoined_wAnnotation)
# 'data.frame':	41253 obs. of  15 variables:
#   $ Element_Name                 : chr  "A_51_P470981" "A_51_P269703" "A_52_P529098" "A_52_P214058" ...
# $ Gene_Symbol                  : logi  NA NA NA NA NA NA ...
# $ Gene_Name                    : logi  NA NA NA NA NA NA ...
# $ NCBI_ID                      : logi  NA NA NA NA NA NA ...
# $ FoldChange_24.h              : num  -0.02049 0.0149 0.00917 -0.05741 -0.06237 ...
# $ Tstat_24.h                   : num  -2.115 1.33 0.761 -6.42 -4.13 ...
# $ PValue_24.h                  : num  0.06739 0.2203 0.4683 0.000205 0.003297 ...
# $ Gene_Symbol                  : logi  NA NA NA NA NA NA ...
# $ Gene_Name                    : logi  NA NA NA NA NA NA ...
# $ NCBI_ID                      : logi  NA NA NA NA NA NA ...
# $ FoldChange_lipopolysaccharide: num  -0.0441 0.0192 0.0194 -0.0578 0.1146 ...
# $ Tstat_lipopolysaccharide     : num  -4.55 1.72 1.61 -6.46 7.59 ...
# $ PValue_lipopolysaccharide    : num  1.87e-03 1.24e-01 1.46e-01 1.96e-04 6.37e-05 ...
# $ Gene_Symbol_New              : chr  "Fut10" "Klra2" "1600012F09Rik" "Raf1" ...
# $ Gene_Name_New                : chr  "fucosyltransferase 10" "killer cell lectin-like receptor, subfamily A, member 2" "RIKEN cDNA 1600012F09 gene" "v-raf-leukemia viral oncogene 1" ...

#Then, to make the lay out for our differential expression results data frame similar to what we used for our other datasets, I put the new gene symbol annotation into the old gene symbol annotation column, and then do the same for the new gene name annotation:
TempResultsJoined_wAnnotation$Gene_Symbol<-TempResultsJoined_wAnnotation$Gene_Symbol_New
TempResultsJoined_wAnnotation$Gene_Name<-TempResultsJoined_wAnnotation$Gene_Name_New

str(TempResultsJoined_wAnnotation)
# 'data.frame':	41253 obs. of  15 variables:
#   $ Element_Name                 : chr  "A_51_P470981" "A_51_P269703" "A_52_P529098" "A_52_P214058" ...
# $ Gene_Symbol                  : chr  "Fut10" "Klra2" "1600012F09Rik" "Raf1" ...
# $ Gene_Name                    : chr  "fucosyltransferase 10" "killer cell lectin-like receptor, subfamily A, member 2" "RIKEN cDNA 1600012F09 gene" "v-raf-leukemia viral oncogene 1" ...
# $ NCBI_ID                      : logi  NA NA NA NA NA NA ...
# $ FoldChange_24.h              : num  -0.02049 0.0149 0.00917 -0.05741 -0.06237 ...
# $ Tstat_24.h                   : num  -2.115 1.33 0.761 -6.42 -4.13 ...
# $ PValue_24.h                  : num  0.06739 0.2203 0.4683 0.000205 0.003297 ...
# $ Gene_Symbol                  : logi  NA NA NA NA NA NA ...
# $ Gene_Name                    : logi  NA NA NA NA NA NA ...
# $ NCBI_ID                      : logi  NA NA NA NA NA NA ...
# $ FoldChange_lipopolysaccharide: num  -0.0441 0.0192 0.0194 -0.0578 0.1146 ...
# $ Tstat_lipopolysaccharide     : num  -4.55 1.72 1.61 -6.46 7.59 ...
# $ PValue_lipopolysaccharide    : num  1.87e-03 1.24e-01 1.46e-01 1.96e-04 6.37e-05 ...
# $ Gene_Symbol_New              : chr  "Fut10" "Klra2" "1600012F09Rik" "Raf1" ...
# $ Gene_Name_New                : chr  "fucosyltransferase 10" "killer cell lectin-like receptor, subfamily A, member 2" "RIKEN cDNA 1600012F09 gene" "v-raf-leukemia viral oncogene 1" ...

#Then I overwrite the original TempResultsJoined object so that we can just feed it into our normal code pipeline:
TempResultsJoined<-TempResultsJoined_wAnnotation

#And then the rest of the coding should be the same as normal:

FilteringDEResults_GoodAnnotation(TempResultsJoined)

# [1] "# of rows with missing NCBI annotation:"
# [1] NA
# [1] "# of rows with missing Gene Symbol annotation:"
# [1] 6360
# [1] "# of rows mapped to multiple NCBI_IDs:"
# [1] 0
# [1] "# of rows mapped to multiple Gene Symbols:"
# [1] 0
# [1] "# of rows with good annotation"
# [1] 34893
# [1] "Outputted object: TempResultsJoined_NoNA_NoMultimapped"
# Warning message:
#   In rm(TempResultsJoined_NoNA, TempResultsJoined_NoNA_NoMultimapped) :
#   object 'TempResultsJoined_NoNA_NoMultimapped' not found

str(TempResultsJoined_NoNA_NoMultimapped)
# 'data.frame':	34893 obs. of  15 variables:
#   $ Element_Name                 : chr  "A_51_P470981" "A_51_P269703" "A_52_P529098" "A_52_P214058" ...
# $ Gene_Symbol                  : chr  "Fut10" "Klra2" "1600012F09Rik" "Raf1" ...
# $ Gene_Name                    : chr  "fucosyltransferase 10" "killer cell lectin-like receptor, subfamily A, member 2" "RIKEN cDNA 1600012F09 gene" "v-raf-leukemia viral oncogene 1" ...
# $ NCBI_ID                      : logi  NA NA NA NA NA NA ...
# $ FoldChange_24.h              : num  -0.02049 0.0149 0.00917 -0.05741 -0.06237 ...
# $ Tstat_24.h                   : num  -2.115 1.33 0.761 -6.42 -4.13 ...
# $ PValue_24.h                  : num  0.06739 0.2203 0.4683 0.000205 0.003297 ...
# $ Gene_Symbol                  : logi  NA NA NA NA NA NA ...
# $ Gene_Name                    : logi  NA NA NA NA NA NA ...
# $ NCBI_ID                      : logi  NA NA NA NA NA NA ...
# $ FoldChange_lipopolysaccharide: num  -0.0441 0.0192 0.0194 -0.0578 0.1146 ...
# $ Tstat_lipopolysaccharide     : num  -4.55 1.72 1.61 -6.46 7.59 ...
# $ PValue_lipopolysaccharide    : num  1.87e-03 1.24e-01 1.46e-01 1.96e-04 6.37e-05 ...
# $ Gene_Symbol_New              : chr  "Fut10" "Klra2" "1600012F09Rik" "Raf1" ...
# $ Gene_Name_New                : chr  "fucosyltransferase 10" "killer cell lectin-like receptor, subfamily A, member 2" "RIKEN cDNA 1600012F09 gene" "v-raf-leukemia viral oncogene 1" ...

CollapsingDEResults_OneResultPerGene(GSE_ID="GSE81024", TempResultsJoined_NoNA_NoMultimapped, ComparisonsOfInterest = c("GSE81024_LPS_vs_Ctrl"), NamesOfFoldChangeColumns=list(TempResultsJoined_NoNA_NoMultimapped$FoldChange_lipopolysaccharide), NamesOfTstatColumns = list(TempResultsJoined_NoNA_NoMultimapped$Tstat_lipopolysaccharide))

# [1] "Double check that the vectors containing the two fold change and tstat column names contain the same order as the group comparisons of interest - otherwise this function won't work properly!  If the order matches, proceed:"
# [1] "# of rows with unique NCBI IDs:"
# [1] 1
# [1] "# of rows with unique Gene Symbols:"
# [1] 21192
# [1] "Dimensions of Fold Change matrix, averaged by gene symbol:"
# [1] 21192     1
# [1] "Output: Named DEResults_GSE81024"

str(DEResults_GSE81024)
# List of 4
# $ Log2FC: num [1:21192, 1] -0.02381 -0.01922 -0.07282 0.01491 0.00105 ...
# ..- attr(*, "dimnames")=List of 2
# .. ..$ : chr [1:21192] "0610005C13Rik" "0610007C21Rik" "0610007L01Rik" "0610007N19Rik" ...
# .. ..$ : chr "GSE81024_LPS_vs_Ctrl"
# $ Tstat : num [1:21192, 1] -1.988 -2.905 -10.332 1.183 0.121 ...
# ..- attr(*, "dimnames")=List of 2
# .. ..$ : chr [1:21192] "0610005C13Rik" "0610007C21Rik" "0610007L01Rik" "0610007N19Rik" ...
# .. ..$ : chr "GSE81024_LPS_vs_Ctrl"
# $ SE    : num [1:21192, 1] 0.01198 0.00662 0.00776 0.0126 0.01533 ...
# ..- attr(*, "dimnames")=List of 2
# .. ..$ : chr [1:21192] "0610005C13Rik" "0610007C21Rik" "0610007L01Rik" "0610007N19Rik" ...
# .. ..$ : chr "GSE81024_LPS_vs_Ctrl"
# $ SV    : num [1:21192, 1] 1.43e-04 4.38e-05 6.03e-05 1.59e-04 2.35e-04 ...
# ..- attr(*, "dimnames")=List of 2
# .. ..$ : chr [1:21192] "0610005C13Rik" "0610007C21Rik" "0610007L01Rik" "0610007N19Rik" ...
# .. ..$ : chr "GSE81024_LPS_vs_Ctrl"

#Done :)

###########################
